File: set/src/client/PlayerClient.java
package src.client;

import src.action.Action;
import src.action.ActionType;
import src.action.PlayerAction;
import src.action.PlayerActionSelectSet;
import src.action.actionQueue.SynchronisedActionQueue;
import src.networkHelpers.SocketReader;
import src.networkHelpers.SocketWriter;
import src.player.PlayerInteractor;

import java.io.BufferedWriter;
import java.io.IOException;
import java.io.OutputStreamWriter;
import java.net.InetAddress;
import java.net.Socket;
import java.util.List;
import java.util.Optional;

import static src.action.ActionType.LEAVE_GAME;
import static src.player.PlayerInteractor.farewellPlayer;
import static src.player.PlayerInteractor.readPlayerName;


public class PlayerClient {
    private final static int GAME_PORT = 9090;
    private final static SynchronisedActionQueue actions = new SynchronisedActionQueue();

    public static void main(String[] args) throws IOException, InterruptedException {

        /*
        List<Integer> cardIds = List.of(42, 69, 420);
        ActionProtos.Action protoAction = ActionProtos.Action.newBuilder()
            .setType(ActionProtos.ActionType.SELECT_SET)
            .setPlayerID(1337)
            .addAllCardIDs(cardIds)
            .build();
        System.out.println("raw proto");
        System.out.println(protoAction);
        System.out.println("raw proto to string");
        System.out.println(protoAction.toString());
        System.out.println("serialised proto");
        System.out.println(protoAction.toByteArray());
        try (FileOutputStream stream = new FileOutputStream("proto_out")) {
            stream.write(protoAction.toByteArray());
        }*/

//        try (FileInputStream stream = new FileInputStream("proto_out")) {
//            byte[] serializedProto = stream.readAllBytes();
//            ActionProtos.Action protoAction = ActionProtos.Action.parseFrom(serializedProto);
//            System.out.println("from bytes");
//            System.out.println(protoAction);
//        }



        Optional<Integer> playerID = Optional.empty();
        SocketReader fromServer;
        SocketWriter toServer;

        try {
            InetAddress host = InetAddress.getLocalHost();
            Socket fromServerSocket = new Socket(host.getHostName(), GAME_PORT);
            fromServer = new SocketReader(fromServerSocket);

            // once off (just to establish a connection) we will write to the thing that will read
            BufferedWriter writer = null;
            try {
                writer = new BufferedWriter(new OutputStreamWriter(fromServerSocket.getOutputStream()));
                String playerName = readPlayerName();
                System.out.println("REQUEST_PLAYER_ID");
                String joinGameRequestMsg = "REQUEST_PLAYER_ID" + "@" + playerName + "\n";
                writer.write(joinGameRequestMsg);
                writer.flush();

                // exponential backoff yay
                long delay = 25;
                while (!playerID.isPresent()) {
                    String msg = fromServer.readLine().toString();
                    System.out.println(msg);
                    if (msg.equals("REQUEST_PLAYER_ID")) {
                        System.out.println("TRYING TO GET PLAYER ID");
                        writer.write(joinGameRequestMsg);
                        writer.flush();
                        Thread.sleep(delay);
                        delay *= 2;
                    } else playerID = Optional.of(Integer.valueOf(msg));
                }
            } catch (Exception e) {
                if (writer != null) writer.close();
            }

            try {
                toServer = new SocketWriter(new Socket(host.getHostName(), GAME_PORT));
                toServer.write(playerID.get());
            } catch (IOException e) {
                fromServer.close();
                throw e;
            }
        } catch (Exception e) {
            System.out.println(e.getMessage());
            throw e;
        }

        Thread actionDrainer = new Thread(new ActionQueueDrainer(actions, toServer));
        actionDrainer.start();

        Thread feedbackGetter = new Thread(new FeedbackGetter(fromServer));
        feedbackGetter.start();

        // deals with stdin moves
        while (true) {
            PlayerAction action = PlayerInteractor.getAction(playerID.get());
            if (action != null) {
                actions.addAction(action);
                System.out.println("adding action to playerqueue");

                if (action.getType() == LEAVE_GAME) {
                    farewellPlayer();
                    break;
                }
            }
        }
        // TODO: call this when above loop ain't infinite t.join();
        // TODO: close() (my function close) reader and writer and
    }
}

// thread that gets actions by scanning stdin and adds to actions queue
class ActionQueueDrainer implements Runnable {
    private final SocketWriter toServer;
    private final SynchronisedActionQueue actions;

    public ActionQueueDrainer(SynchronisedActionQueue actions, SocketWriter toServer) {
        this.actions = actions;
        this.toServer = toServer;
    }

    @Override
    public void run() {
// other thread sends a move to server as soon as there is one
        while (true) {
            Action action = actions.getNext();

            assert(action != null);

            try {
                if (action != null) {
                    System.out.println("client is sending an action!");
                    sendAction(action);
                }
            } catch (IOException e) {
                // TODO: handle this exception
                 continue;
            }

        }
    }

    private void sendAction(Action action) throws IOException {
        toServer.write(serializeAction((PlayerAction) action));
    }

    // moveType@playerID@optionalSetCardIDsSeparatedByAtSeparator
    // for example "SELECT_SET@1@2@4@16"
    // for example "REQUEST_DRAW_3@3"
    private String serializeAction(PlayerAction action) {
        ActionType actionType = action.getType();
        String serializedAction = actionType + "@" + action.getPlayerID();

        if (actionType == ActionType.SELECT_SET) {
            List<Integer> cardPositions = ((PlayerActionSelectSet) action).getCardPositions();
            for (int cardPosition: cardPositions) {
                serializedAction += "@" + cardPosition;
            }
        }
        System.out.println(serializedAction);
        return serializedAction;
    }
}

// This thread is in charge of displaying it the game may send a message to a player.
class FeedbackGetter implements Runnable {
    private final SocketReader fromServer;

    public FeedbackGetter(SocketReader fromServer) {
        this.fromServer = fromServer;
    }

    @Override
    public void run() {
// other thread sends a move to server as soon as there is one
        while (true) {
            try {
                System.out.println(fromServer.readLine());
            } catch (IOException e) {
                System.out.println(e.getMessage());
            }
        }
    }
}

File: set/src/client/PlayerClient.py
# Play Set
# Dylan Blecher
# blecher.dylan@gmail.com
# April-June 2020
# Run player client

from queue import Queue
from socket import gethostname
from enum import Enum
from client.Structures import (
    ActionType,
    PlayerAction,
    PlayerActionSelectSet,
)

from proto.action_pb2 import (
    ClientRequest as ClientRequestProto,
    ServerResponse as ServerResponseProto,
    JoinGameRequest as JoinGameRequestProto,
    ConfirmPlayerID as ConfirmPlayerIDProto
)

import socket
import time
import threading
import json

from client import server_communicator

from google.protobuf.internal.encoder import _VarintEncoder
from google.protobuf.json_format import MessageToJson

from flask import render_template


SET_SIZE = 3
GAME_PORT = 9090        # The port used by the server

# Run the player 
# (setup socket connection and then all interaction with Java server and Javascript front end)
class PlayerRunner:
    def __init__(self, sock):
        self._sock = sock
        self._actions = Queue() # python queues are synchronous if you add a few flags when using :)
        self.playerID = None
    
    def run_player(self):    
        # setup socke connection with backend Java
        # HOST = 'Dylans-MacBook-Pro-2.local'
        HOST = socket.gethostbyname('localhost')
        # HOST = gethostname()    # The server's hostname or IP address

        print("HOST IS: " + HOST)
        fromServer = None
        toServer = None
        
        try:
            fromServer = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
            fromServer.connect((HOST, GAME_PORT))

            playerIDRequestProto = ClientRequestProto(
                joinGame=JoinGameRequestProto(
                    name= "Dyl" # readPlayerName()
                )
            )

            server_communicator.send_message(fromServer, playerIDRequestProto)


            delay = 0.025 # 25 milliseconds
            while self.playerID is None:
                time.sleep(delay)
                delay *= 2
                
                try: 
                    serverResponseProto = server_communicator.receive_message(fromServer, ServerResponseProto)
                    # print(serverResponseProto)
                    if serverResponseProto.HasField("playerID"):
                        self.playerID = serverResponseProto.playerID
                        
                        # can't write to fromServer anymore - it's one way!
                        fromServer.shutdown(socket.SHUT_WR) 

                except Exception as e:
                    print(str(e))
                    server_communicator.send_message(fromServer, playerIDRequestProto)
            
            toServer = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
            toServer.connect((HOST, GAME_PORT))  
            
            # can't read from toServer anymore - it's one way!
            toServer.shutdown(socket.SHUT_RD) 

            confirmPlayerIDProto = ClientRequestProto(
                confirmPlayerID=ConfirmPlayerIDProto(
                    playerID=self.playerID
                )
            )
            
            server_communicator.send_message(toServer, confirmPlayerIDProto)
        
        except Exception as e:
            print(str(e))
            if fromServer is None:
                closeSocket(toServer)
            if toServer is None:
                closeSocket(fromServer)
            raise e

        actionQueueThread = self._sock.start_background_task(ActionQueueDrainer(self._actions, toServer).run)
        feedbackThread = self._sock.start_background_task(FeedbackGetter(fromServer, self._sock).run)

        self._sock.sleep(5)
        actionQueueThread.join()
        feedbackThread.join()

    def addActionToQueue(self, action):
        self._actions.put(action)

    def getPlayerID(self):
        return self.playerID

def closeSocket(sock):
    if sock != None:
        sock.shutdown(socket.SHUT_RDWR)
        sock.close()

# Get an action and send it to the server
class ActionQueueDrainer:
    def __init__(self, actions, toServer):
        self.actions = actions
        self.toServer = toServer

    def run(self):
        while True:
            action = self.actions.get(block=True)
            assert(action != None)
            self.sendAction(action)

    def sendAction(self, action: PlayerAction) -> None:
        server_communicator.send_message(self.toServer, action.proto)

# Handle messages from the server
class FeedbackGetter:
    def __init__(self, fromServer, sock):
        self.fromServer = fromServer
        self._sock = sock

    def run(self):
        while True:
            try:
                # CHANEG FROM _PRIVATE_METHOD
                serverResponseProto = server_communicator.receive_message(self.fromServer, ServerResponseProto)
                # print("got message" + str(serverResponseProto))
                if serverResponseProto.HasField("errorMessage"):
                    self._sock.emit("error_message", str(serverResponseProto.errorMessage))
                elif serverResponseProto.HasField("state"):
                    print("got state")
                    # convert state to json
                    stateJson = MessageToJson(serverResponseProto.state, indent=None)
                    self._sock.emit("board_change", stateJson)
                elif serverResponseProto.HasField("revealedSet"):
                    revealSetJson = MessageToJson(serverResponseProto.revealedSet, indent=None)
                    self._sock.emit("reveal_set", revealSetJson)
                    time.sleep(2.5) # TODO: This sleep is way too hacky. It should all be front end.
                    # Once I have the program running based on sending actions, instead of state
                    # every loop of the game-runner, this won't be an issue.
                    # The current issue that this sleep fixes is that it's overwriting the set 
                    # cards too quickly since state is sent straight after the revealed set is sent.
                elif serverResponseProto.HasField("result"):
                    break
            except Exception as e:
                print(str(e))